rllib/core/learner/training_data.py

```diff
@@ -0,0 +1,144 @@
+from collections import defaultdict
+import dataclasses
+from typing import List, Optional
+
+import tree  # pip install dm_tree
+
+import ray
+from ray.rllib.env.multi_agent_episode import MultiAgentEpisode
+from ray.rllib.policy.sample_batch import MultiAgentBatch
+from ray.rllib.utils.minibatch_utils import (
+    ShardBatchIterator,
+    ShardEpisodesIterator,
+    ShardObjectRefIterator,
+)
+from ray.rllib.utils.typing import EpisodeType
+
+
+@dataclasses.dataclass
```

Comment:
As this class might be created quite often, implementing it with `@dataclasses.dataclass(slots=true)` would make it more memory-efficient.

Reply:
Great catch. Changed.